BDArmory/WeaponMounts/ModuleCustomTurret.cs

```diff
@@ -0,0 +1,248 @@
+                if (!yawTransform)
+                {
+                    Debug.LogWarning("[BDArmory.ModuleCustomTurret]: " + part.partInfo.title + " has no yawTransform");
+                }
+                if (!bottomTransform)
+                {
+                    Debug.LogWarning("[BDArmory.ModuleCustomTurret]: " + part.partInfo.title + " has no bottomTransform");
+                }
+            }
+            if (!referenceTransform)
+            {
+                if (pitchTransform)
+                    SetReferenceTransform(pitchTransform);
+                else if (yawTransform)
+                    SetReferenceTransform(yawTransform);
+                else
+                    Debug.LogWarning("[BDArmory.ModuleCustomTurret]: " + part.partInfo.title + " has no referenceTransform");
+            }
+            if (!bottomTransform) bottomTransform = part.transform;
+
+            yawNormal = yawTransform.up;
+            //because ofc Squad can't have consistant standard for servo/hinge axis transform orientation...
+            //Also need to account for rotation/facing; ModuleTurret is agnostic, but targetAngle in the hinge module is not.
+            if (Hinge)
+            {
+                if (Hinge.mainAxis == "X") yawNormal = -pitchTransform.forward;
+                if (Hinge.mainAxis == "Z") yawNormal = pitchTransform.right;
+            }
+        }
+
+        public void AimToTarget(Vector3 targetPosition, bool pitch = true, bool yaw = true)
+        {
+            AimInDirection(targetPosition - referenceTransform.position);
+        }
+
+        public void AimInDirection(Vector3 targetDirection)
+        {
+            if ((Servo && !yawTransform) || (Hinge && !pitchTransform))
+            {
+                return;
+            }
+            if (!bottomTransform) return;
+
+            Vector3 yawComponent = targetDirection.ProjectOnPlanePreNormalized(yawNormal);
+            Vector3 pitchComponent = targetDirection.ProjectOnPlane(Vector3.Cross(yawComponent, yawNormal));
+            //float currentYaw = Hinge ? 0 : Servo ? Servo.currentAngle : 0; //currentAngle for whatever reason only updates when the PAW is open. WTF, KSP.
+            float currentYaw = Servo ? VectorUtils.SignedAngleDP(bottomTransform.forward, referenceTransform.forward, bottomTransform.right) : 0;
+            float yawError = VectorUtils.SignedAngleDP(
+                referenceTransform.forward.ProjectOnPlanePreNormalized(yawNormal),
+                yawComponent,
+                Vector3.Cross(yawNormal, referenceTransform.forward));
+            float targetYawAngle = (currentYaw + yawError).ToAngle();
+            // clamp target yaw in a non-wobbly way
+            if (fullRotation)
+            {
+                if (Mathf.Abs(targetYawAngle) > 180)
+                {
+                    var nonWooblyWay = Vector3.Dot(yawTransform.parent.right, targetDirection + referenceTransform.position - yawTransform.position);
```

Comment:
While `nonWooblyWay` sounds funny, I suspect it ought to be `nonWobblyWay`...

Also, any idea why this needs to check for being `NaN`? I don't see this happening unless the Kraken has already struck and broken one of the transforms, in which case, the various `NaN` checks could come earlier.

Reply:
It's like that in ModuleTurret too.

Reply:
Yeah, I noticed that.
Any idea why the `NaN` checks are needed? The `if (float.IsNaN(currentYaw))` check should probably be moved to immediately after where `currentYaw` is calculated if it's actually needed.